### PR TITLE
Feature quay integration

### PR DESCRIPTION
**Pull request by little-arhat**
### How to use admin interface

Homu provides admin interface to add and auto-register repos in github and specified builder. Currently
homu can only autoregister repo in Quay.io.

To add repo to homu and register it within github & quay:

``` sh
$ curl -XPUT -H"Authorization: {ADMUSER} {ADMTOKEN}" -H"Content-type: application/json" -d'{"owner": "{gh_owner}", "name": "{gh_repo}", "reviewers": ["{gh_reviewer}"], "github": {}, "builder": "quay", "quay": {"public": true}}' "http://{HOMU_IP}:{HOMU_PORT}/admin/repo"'
{"quay": {"secret": "{secr}", "username": "{user}", "name": "{gh_repo}", "build_branches": ["master", "develop", "auto"], "ssh": {deploy_key_id}, "public": true, "url": "{quay_repo_url}", "webhooks": "{quay_webhook}"}, "owner": "{gh_owner}", "reviewers": ["{gh_reviewer}"], "builder": "quay", "name": "{gh_repo}", "github": {"secret": "{gh_secret}", "webhook_id": {gh_webhook_id}}, "label": "{repo_label}"}
```

This command will add repo to homu, store it to db for persistence, create hook on github, create repo
on Quay.io, and register all necessary hooks there. One can add repos using other builders, but autoregistration is not supported for them.

If you want to register existing quay repo in homu, add `"existing_repo"` key to `"quay"` settings and set it to name of your quay repo.

To delete repo and unregister it send DELETE request to `admind/repo/{repo_label}`. To keep quay repo undeleted after unregistering, pass `?keep_repo=True`.

``` sh
$ curl -XDELETE -H"Authorization: {ADMUSER} {ADMTOKEN}" "http://{homu_ip}:{homu_port}/admin/repo/{repo_label}"
{repo_label}
```

For builders other than Quay.io auto unregistration is not supported.

### CODE REVIEW

File: homu/quay1.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+QUAY_API = 'api/v1/'
+CSRF_TOKEN_RE = re.compile(r'__token\s?=\s?\'(.*)\';')
+
+EVENT_BUILD_SUCCESS = 'build_success'
+EVENT_BUILD_FAILURE = 'build_failure'
+
+# import http.client as http_client
+# http_client.HTTPConnection.debuglevel = 1
+
+class QuayThing:
```
Comment (regadas) [inline homu/quay1.py:22]: why thing? :P (is it by chance a reference to the movie "The Thing")

File: homu/quay1.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        info = r.json()
+        return self.repo(info['name'])
+
+    # Get the repo object (NOTE: no requests are made here)
+    def repo(self, name):
+        return self._create_child(Repo, name)
+
+class Repo(NamespacedThing):
+    def __init__(self, name, **kwargs):
+        super(self.__class__, self).__init__(**kwargs)
```
Comment (regadas) [inline homu/quay1.py]: warn: using self.**class** in super might result in something like RuntimeError: maximum recursion

File: homu/utils.py [hunk 3]
```diff
@@ -46,3 +52,85 @@ def remove_url_keys_from_json(json):
...
+    except error as e:
+        if logger:
+            logger.warning(e)
+
+def update_in(dkt, key, func):
+    if key in dkt:
+        dkt[key] = func(dkt[key])
+    return dkt
+
+def merge_dicts(fst, snd=None, **kwargs):
```
Comment (regadas) [inline homu/utils.py]: Alternative solution to chain several stuff `merge_dicts(dict(asd=1), dict(qwe=2), asd=1, qwe=3)` might save some lines 

```
from itertools import chain

def merge_dicts(*args, **kwargs):
    merged = chain(
        chain(*(d.items() for d in args)),
        kwargs.items()
    )

    return dict(merged)
```

**Comment by regadas** [comment]
I'm going to merge this ... Gonna add docker support also

File: homu/quay1.py (hunk 2 above)
Comment (little-arhat) [inline homu/quay1.py]: forgot that's python3, so we can just use `super().__init__` %)

File: homu/utils.py (hunk 3 above)
Comment (little-arhat) [inline homu/utils.py]: Agreed, that's more flexible.

File: homu/quay1.py (hunk 1 above)
Comment (little-arhat) [inline homu/quay1.py:22]: %) no, that's like `Entity` or `Object`. `github3` uses `GitHubObject` as name for base class that `Repo`, `Hook`, etc extends. We can use that to or stick with "The Thing" %)

**Pull request merged**